### PR TITLE
Split cloud-aws into repository-s3 and discovery-ec2

### PR DESCRIPTION
**Pull request by dadoonet**
Until now we had a cloud-aws plugin which is providing 2 disctinct features:
- discovery on EC2
- snapshot/restore on S3

This commit splits the plugin by feature so people can use either one or the other or both features.

Doc is updated accordingly.

If we do this split, we will also need to add new github labels `:Plugin Repository S3` and `:Plugin Discovery EC2`.

### CODE REVIEW

File: plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/DiscoveryEc2Plugin.java [hunk 1]
```diff
@@ -37,49 +33,32 @@
...
     @Override
     public String name() {
-        return "cloud-aws";
+        return "discovery-ec2";
     }
 
     @Override
     public String description() {
-        return "Cloud AWS Plugin";
+        return "Discovery EC2 Plugin";
```
Comment (rjernst) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/DiscoveryEc2Plugin.java]: While I think the name in maven should keep with prepending the plugin type, it makes more sense gramatically to say "EC2 Discovery Plugin" in the description (and this is what I did for multicast).

Comment (rjernst) [inline plugins/discovery-ec2/src/main/java/org/elasticsearch/plugin/discovery/ec2/DiscoveryEc2Plugin.java]: I also did this for class names for the same reason.

**Comment by dadoonet** [comment]
@rjernst Thanks for the comment. Yeah it makes more sense to me. And BTW it's conform to the documentation titles I added :) 

PR updated for discovery and repository

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 2]
```diff
@@ -80,15 +80,16 @@
                     "analysis-phonetic",
                     "analysis-smartcn",
                     "analysis-stempel",
-                    "cloud-aws",
                     "cloud-azure",
                     "cloud-gce",
                     "delete-by-query",
                     "discovery-multicast",
+                    "discovery-ec2",
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: I like the fact that it's alphabetically sorted :) so let's put it before multicast?

File: docs/plugins/cloud-aws.asciidoc [hunk 3]
```diff
@@ -1,468 +1,8 @@
 [[cloud-aws]]
 === AWS Cloud Plugin
 
-The Amazon Web Service (AWS) Cloud plugin uses the
-https://github.com/aws/aws-sdk-java[AWS API] for unicast discovery, and adds
-support for using S3 as a repository for
-{ref}/modules-snapshots.html[Snapshot/Restore].
+The AWS Cloud Plugin has been split in two plugins:
```
Comment (jpountz) [inline docs/plugins/cloud-aws.asciidoc]: I think this page should be removed and added to redirects.asciidoc?

File: docs/plugins/discovery.asciidoc [hunk 4]
```diff
@@ -47,5 +44,9 @@ include::cloud-gce.asciidoc[]
 
 include::discovery-multicast.asciidoc[]
 
+////
+Deprecated pages
+////
+include::cloud-aws.asciidoc[]
```
Comment (jpountz) [inline docs/plugins/discovery.asciidoc]: remove and add redirects?

File: plugins/pom.xml [hunk 5]
```diff
@@ -432,13 +432,14 @@
         <module>analysis-icu</module>
         <module>cloud-gce</module>
         <module>cloud-azure</module>
-        <module>cloud-aws</module>
         <module>delete-by-query</module>
         <module>discovery-multicast</module>
+        <module>discovery-ec2</module>
```
Comment (jpountz) [inline plugins/pom.xml]: before multicast?

File: plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+        Collection<Module> modules = new ArrayList<>();
+        modules.add(new S3Module());
+        return modules;
+    }
+
+    @Override
+    public Collection<Class<? extends LifecycleComponent>> nodeServices() {
+        Collection<Class<? extends LifecycleComponent>> services = new ArrayList<>();
+        services.add(S3Module.getS3ServiceImpl());
+        return services;
```
Comment (jpountz) [inline plugins/repository-s3/src/main/java/org/elasticsearch/plugin/repository/s3/S3RepositoryPlugin.java]: or just return Collections.singleton(S3Module.getS3ServiceImpl()) ?

File: docs/plugins/cloud-aws.asciidoc (hunk 3 above)
Comment (dadoonet) [inline docs/plugins/cloud-aws.asciidoc]: @clintongormley If I create a new `redirects.asciidoc` file within plugins documentation, will that work?
Or do I need to define something else?

Comment (clintongormley) [inline docs/plugins/cloud-aws.asciidoc]: Given that these docs have never been "current" (they have only recently been added), I wouldn't worry about the redirects.  Just make the changes directly.

Comment (dadoonet) [inline docs/plugins/cloud-aws.asciidoc]: Indeed. It makes sense!

**Comment by dadoonet** [comment]
@jpountz I addressed your comments (and thanks BTW :) )

**Comment by jpountz** [comment]
LGTM. But this probably deserves a second pair of eyes as I am not familiar with this part of the code, so I could easily miss things.

**Comment by dadoonet** [comment]
Thank you @jpountz. @rjernst wanna give a final look?

Then I'll do the same with gce and move it as a discovery plugin (basically renaming only) and azure which needs to be split as well in 2 plugins.

File: docs/plugins/discovery-ec2.asciidoc [hunk 7]
```diff
@@ -0,0 +1,215 @@
...
+==== Removal
+
+The plugin can be removed with the following command:
+
+[source,sh]
+----------------------------------------------------------------
+sudo bin/plugin remove discovery-ec2
+----------------------------------------------------------------
+
+The node must be stopped before removing the plugin.
```
Comment (rjernst) [inline docs/plugins/discovery-ec2.asciidoc]: Not critical for this PR, but do we mention this somewhere else in general plugin docs? It is not possible for any plugin to be added/removed on a running node, so it seems redundant to mention it in the docs for every plugin.

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/Ec2DiscoveryITest.java [hunk 8]
```diff
@@ -40,7 +40,7 @@
     @Test
     public void testStart() {
         Settings nodeSettings = settingsBuilder()
-                .put("plugin.types", CloudAwsPlugin.class.getName())
+                .put("plugin.types", Ec2DiscoveryPlugin.class.getName())
                 .put("cloud.enabled", true)
```
Comment (rjernst) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/Ec2DiscoveryITest.java]: All of these cloud.enabled in tests can be removed right?

**Comment by rjernst** [comment]
LGTM, I left a couple minor comments.

File: plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/Ec2DiscoveryITest.java (hunk 8 above)
Comment (dadoonet) [inline plugins/discovery-ec2/src/test/java/org/elasticsearch/discovery/ec2/Ec2DiscoveryITest.java]: ++ nice catch!

File: docs/plugins/discovery-ec2.asciidoc (hunk 7 above)
Comment (dadoonet) [inline docs/plugins/discovery-ec2.asciidoc]: @clintongormley WDYT? ^^^ This has been added for each plugin.

I'm +1 for keeping this information here even if it's redundant.

Comment (clintongormley) [inline docs/plugins/discovery-ec2.asciidoc]: @rjernst site plugins can be added/removed without restarting.

@dadoonet i'm good with keeping the redundant docs

**Pull request merged**